ecosystem/l2tol2cdm-relayedmessage-return.md

```diff
@@ -0,0 +1,105 @@
+
+## Summary
+
+With an added field to the `RelayedMessage` event emitted by the `L2ToL2CrossDomainMessenger`, we open the door to better async contract development within the Superchain.
+
+## Problem Statement + Context
+
+As we horizontally scale, contract development will look more asynchronous. A lot of cross chain messages are predominalty write-based.
+
+- Bridge Funds
+- Invoke calldata on a remotely controlled proxy account
+
+Read-based cross chain messages are less common but will increasingly become so in an asynchronous world. Solidity has a `view` modifier specifically targeted for reading -- it would be natural to also want to also query this state onchain very quickly.
+
+- Fetch remote swap quotes
+- Query balance/ownership status
+- Chain on subsequent write actions, only after success (deposit -> borrow -> bridge)
+
+In order to support use cases like these, custom handler contracts per-application must be written to capture these side effects and return data backwards or chain on additional side effects. For example
+
+```solidity
+contract ERC20BalanceGateway {
+    L2ToL2CrossDomainMessenger messenger;
+
+    // Gateway contract that can be used to query ERC20s with a specified handler
+    function query(IERC20 token, address account, bytes4 selector) external {
+        uint256 balance = token.balanceOf(account);
+
+        // Message sent back with the value as an encoded argument
+        (address sender, uint256 source) = messenger.crossDomainMessageContext();
+        messenger.sendMessage(source, sender, abi.encode(selector, balance));
+    }
+}
+```
+
+## Proposed Solution
+
+When the `L2ToL2CrossDomainMessenger` invokes a target with calldata, the return data of that call is captured and returned via `relayMessage`. However this return data is only actionable for the caller of this function, typically the relayer, or a faciliating smart contract.
+
+`relayMessage` emits the `RelayedMessage` event on successful execution which itself can be consumed remotely via the `CrossL2Inbox` predeploy. By including the return data in this event, it becomes immediately visible to the sending chain (and all others in the interop set) without any additional message or handler contract.
+
+```solidity
+contract L2ToL2CrossDomainMessenger {
+    function relayMessage(Identifier calldata _id, bytes calldata _sentMessage) {
+        // include return data
+        (, returnData_) = target.call{ value: msg.value }(message);
+        emit RelayedMessage(source, nonce, messageHash, returnData_);
+    }
+}
+```
+
+The `messageHash` known ahead of time can be correlated with the corresponding `RelayedMessage`. Thus the return value of any read function or return value of a write call becomes immediately actionable for the sender, without requiring special integration.
```

Comment:
In order to read/validate the `RelayedMessage`, the identifier of the `RelayedMessage` is a requirement. 
Wouldn't an integration be required for constructing the identifier  of the `RelayedMessage`  since you cannot calculate the identifier ahead of time?

Reply:
by ahead of time here, I meants with the return value of `sendMessage`. You can directly chain the continuation since the caller just needs to assert the msg hash matches with the `RelayedMessage` event.

I'll make this a bit more clear

Reply:
I was referring to the message identifier not being able to be calculated ahead of time and not the message hash. The identifier is needed in order to trigger the continuation, right?

Reply:
yea after relay is when the continuation would be invoked. you need the return value

the identifier isnt computed ahead of time. It can just be consumed right away after emission

Reply:
Yeah and so that identifier and message will need to be fetched or indexed offchain. Do you see this requiring some additional offchain infra to find and relay these events back to the source chain?

Reply:
yep, we'll definitely need indexing of pending promises that relayers will have to handle separately. Especially now that the data is hashed.

This plays in nicely for protocols that have to propagate the return data backwards, which incurs extra gas cost